File: 0027-remove-element/0027-remove-element.py
from typing import List

class Solution:
    def removeElement(self, nums: List[int], val: int) -> int:
        """
        Removes all occurrences of a specified value from the list in-place.

        Parameters:
        nums (List[int]): List of integers to process.
        val (int): The value to remove from the list.

        Returns:
        int: The new length of the list after removing the specified value.
        """
        # Start iterating from the end of the list
        index = len(nums) - 1

        # Traverse the list backwards to avoid shifting elements unnecessarily
        while index >= 0:
            if nums[index] == val:
                nums.pop(index)  # Remove the element at the current index
            index -= 1  # Decrement the index to check the next element

        # Return the new length of the list
        return len(nums)

File: Explore Card Track/4. Array and String/0027-remove-element/0027-remove-element.c
#include <stdio.h>

/**
 * @brief Removes all occurrences of a specified value from an array in-place.
 *
 * @param nums (int*): Pointer to the array of integers.
 * @param numsSize (int): The size of the array.
 * @param val (int): The value to remove from the array.
 *
 * @return (int): The new length of the array after removing the specified value.
 */
int removeElement(int* nums, int numsSize, int val) {
    int notValIndex = 0;  // Index to place elements not equal to `val`

    // Iterate through the array
    for (int i = 0; i < numsSize; i++) {
        // If the current element is not equal to `val`, move it to `notValIndex`
        if (nums[i] != val) {
            nums[notValIndex] = nums[i];  // Overwrite the position at `notValIndex`
            notValIndex++;               // Increment the index for the next non-val element
        }
    }

    // Return the new size of the array (number of elements not equal to `val`)
    return notValIndex;
}
